### ISSUE TITLE
Add a CEIL rounding mode

### ISSUE THREAD

**Issue by npomfret** [ccxt/ccxt#6645]
I'm trying to use price to precision to round a calculated price.  I want to make sure that prices for ask orders are rounded up, and prices for bid orders are rounded down.  Is this possible in the ccxt api?

**Comment by kroitor** [comment]
> Is this possible in the ccxt api?

Partially, but not entirely. We need to add ROUND_UP and ROUND_DOWN to the set of supported rounding modes for `decimalToPrecision`. More about it here:

- https://github.com/ccxt/ccxt/issues/2733

**Comment by kroitor** [comment]
P.S. We'll do our best to add the support for those two rounding modes as soon as we can – your help would be very appreciated, thx!

**Comment by npomfret** [comment]
Hmm, yes I just found that.  I can't quite get what I'm after.  The problem I have is that I want to round say '0.1419264' to '0.15' of the bitbay ETH/EUR market.  Of course, currently even using ROUND_UP that get's rounded to '0.14'.

**Comment by kroitor** [comment]
The problem I have is that I want to round say '0.1419264' to '0.15' of the bitbay ETH/EUR market.

In general, there's several types of rounding in the family:

- floor (truncate – we have it)
- round (we have it)
- ceil ← we might need to introduce this to round 0.141 to 0.15

**Comment by npomfret** [comment]
Is floor called `TRUNCATE` in this case?

**Comment by kroitor** [comment]
> Is floor called TRUNCATE in this case?

In short, yes, they are identical in almost all aspects.

**Comment by npomfret** [comment]
Is the following an over simplification?

        const input = 0.1419264;
        const precision = 2;
        console.log({input, precision});

        const num1 = input * Math.pow(10, precision);
        console.log({num1});

        const num2 = Math.ceil(num1);// apply rounding here ceil, floor etc
        console.log({num2});

        const num3 = num2 * Math.pow(10, -precision);
        console.log({num3});

->

```
{ input: 0.1419264, precision: 2 }
{ num1: 14.19264 }
{ num2: 15 }
{ num3: 0.15 }
```

**Comment by kroitor** [comment]
> Is the following an over simplification?

It's hard to tell, we usually add the tests for decimals, since they may be very surprising in cases you wouldn't expect... ) It also depends on which counting mode is applied, since precision 2 can mean either of:

- 14.1419264 → 14.15 (decimals "after the dot")
- 14.1419264 → 14.00 (significant digits)

**Comment by npomfret** [comment]
Right.  Are those the only two counting modes?

**Comment by kroitor** [comment]
> Are those the only two counting modes?

They're all here: https://github.com/ccxt/ccxt/blob/master/js/base/functions/number.js#L21

**Comment by npomfret** [comment]
Right, so precision can be lot size or decimal points, which are differentiated by the countingMode  flag?

Is lot size always an integer value?  Or could it be say `0.001`?

**Comment by kroitor** [comment]
@npomfret 

- if countingMode === DECIMAL_PLACES or SIGNIFICANT_DIGITS, precision must contain an integer (for 100, 10, 1, 0.1, 0.01, 0.01 and so on)
- If countingMode === TICK_SIZE, precision must be a float (for 0.001/0.003/0.007/0.013 and so on)

**Comment by npomfret** [comment]
Can we consider this a feature request for a CEIL mode?

**Comment by maartendamen** [comment]
Any news on this, is there already plans for a CEIL mode?

**Comment by kroitor** [comment]
@maartendamen so far we've been able to get away without it in CCXT's basecode. This is not the highest priority, since you can still pre-format your values and do the "ceiling" on the user side. Your contributions are welcome: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md

**Comment by npomfret** [comment]
I've been trying to do my own ceil mode, and have concluded that it's hard!  Does anyone have a good solution that works in each mode?

**Comment by ttodua** [comment]
_(related #2733)_

**Comment by kroitor** [comment]
Hello @npomfret, closing this during a cleanup — this was implemented: the precision module now exports ROUND_UP and ROUND_DOWN rounding modes alongside ROUND and TRUNCATE, which covers the ceil behavior you were building by hand. Thanks for the request!
